core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -410,21 +422,78 @@ public void index(Index index) {
     private void innerIndex(Index index) throws IOException {
         try (Releasable ignored = acquireLock(index.uid())) {
             lastWriteNanos = index.startTime();
-            final long currentVersion;
+            final boolean useAddDocument;
+            /* if we have an autoGeneratedID that comes into the engine we can potentially optimize
+             * and just use addDocument instead of updateDocument and skip the entire version and index lookup across the board.
+             * Yet, we have to deal with multiple document delivery, for this we use a property of the document that is added
+             * to detect if it has potentially been added before. We use the documents timestamp for this since it's something
+             * that:
+             *  - doesn't change per document
+             *  - is preserved in the transaction log
+             *  - and is assigned before we start to index / replicate
+             * NOTE: it's not important for this timestamp to be consistent across nodes etc. it's just a number that is in the common
+             * case increasing and can be used in the failure case when we retry and resent documents to establish a happens before relationship.
+             * for instance:
+             *  - doc A has timestamp 10, autoGeneratedID = true, canHaveDuplicates = true
+             *  - doc B has timestamp 9, autoGeneratedID = true, canHaveDuplicates = true
+             *
+             *  while both docs are in in flight, we disconnect on one node, reconnect and send doc A again
+             *  - now doc A' has timestamp 10, autoGeneratedID = true, canHaveDuplicates = true
+             *
+             *  if A' arrives on the shard first we update highestDeOptimizeAddDocumentTimestamp to 10 and use update document. All subsequent
+             *  documents that arrive (A and B) will also use updateDocument since their timestamps are less than highestDeOptimizeAddDocumentTimestamp.
+             *  While this is not strictly needed for doc B it is just much simpler to implement since it will just de-optimize some doc in the worst case.
+             *
+             *  if A arrives on the shard first we use addDocument since highestDeOptimizeAddDocumentTimestamp is < 10. A` will then just be skipped or calls
+             *  updateDocument.
+             */
+            final boolean skipVersionCheck;
+            if (index.isAutoGeneratedID()
+                && index.version() == Versions.MATCH_ANY
```

Comment:
At second thought, for replicas, the versionType will `EXTERNAL` here, so we can't assert on that either.